From c2ec16a8a9f8c325edc67682368bcfd9c6a50958 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:04:35 -0400
Subject: [powerpc] powernv: Update opal.h to add new LPC and XSCOM functions

Message-id: <1410545655-205645-47-git-send-email-dzickus@redhat.com>
Patchwork-id: 94100
O-Subject: [RHEL7 PATCH 046/626] powerpc/powernv: Update opal.h to add new LPC and XSCOM functions
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit cc0efb57ebf04e066a57aa975c8a38884e603123
Author: Benjamin Herrenschmidt <benh@kernel.crashing.org>
Date:   Mon Jul 15 13:03:09 2013 +1000

    powerpc/powernv: Update opal.h to add new LPC and XSCOM functions

    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/opal.h b/arch/powerpc/include/asm/opal.h
index ea1893b..5d57f6b 100644
--- a/arch/powerpc/include/asm/opal.h
+++ b/arch/powerpc/include/asm/opal.h
@@ -124,6 +124,10 @@ extern int opal_enter_rtas(struct rtas_args *args,
 #define OPAL_PCI_POLL    62
 #define OPAL_PCI_MSI_EOI   63
 #define OPAL_PCI_GET_PHB_DIAG_DATA2  64
+#define OPAL_XSCOM_READ    65
+#define OPAL_XSCOM_WRITE   66
+#define OPAL_LPC_READ    67
+#define OPAL_LPC_WRITE    68
 
 #ifndef __ASSEMBLY__
 
@@ -337,6 +341,17 @@ enum OpalEpowStatus {
  OPAL_EPOW_OVER_INTERNAL_TEMP = 3
 };
 
+/*
+ * Address cycle types for LPC accesses. These also correspond
+ * to the content of the first cell of the "reg" property for
+ * device nodes on the LPC bus
+ */
+enum OpalLPCAddressType {
+ OPAL_LPC_MEM = 0,
+ OPAL_LPC_IO = 1,
+ OPAL_LPC_FW = 2,
+};
+
 struct opal_machine_check_event {
  enum OpalMCE_Version version:8; /* 0x00 */
  uint8_t   in_use;  /* 0x01 */
@@ -697,6 +712,14 @@ int64_t opal_pci_next_error(uint64_t phb_id, uint64_t *first_frozen_pe,
        uint16_t *pci_error_type, uint16_t *severity);
 int64_t opal_pci_poll(uint64_t phb_id);
 
+int64_t opal_xscom_read(uint32_t gcid, uint32_t pcb_addr, uint64_t *val);
+int64_t opal_xscom_write(uint32_t gcid, uint32_t pcb_addr, uint64_t val);
+
+int64_t opal_lpc_write(uint32_t chip_id, enum OpalLPCAddressType addr_type,
+         uint32_t addr, uint32_t data, uint32_t sz);
+int64_t opal_lpc_read(uint32_t chip_id, enum OpalLPCAddressType addr_type,
+        uint32_t addr, uint32_t *data, uint32_t sz);
+
 /* Internal functions */
 extern int early_init_dt_scan_opal(unsigned long node, const char *uname, int depth, void *data);
 
diff --git a/arch/powerpc/platforms/powernv/opal-wrappers.S b/arch/powerpc/platforms/powernv/opal-wrappers.S
index c63791e..c9e67a8 100644
--- a/arch/powerpc/platforms/powernv/opal-wrappers.S
+++ b/arch/powerpc/platforms/powernv/opal-wrappers.S
@@ -118,3 +118,7 @@ OPAL_CALL(opal_pci_next_error,   OPAL_PCI_NEXT_ERROR);
 OPAL_CALL(opal_pci_poll,   OPAL_PCI_POLL);
 OPAL_CALL(opal_pci_msi_eoi,   OPAL_PCI_MSI_EOI);
 OPAL_CALL(opal_pci_get_phb_diag_data2,  OPAL_PCI_GET_PHB_DIAG_DATA2);
+OPAL_CALL(opal_xscom_read,   OPAL_XSCOM_READ);
+OPAL_CALL(opal_xscom_write,   OPAL_XSCOM_WRITE);
+OPAL_CALL(opal_lpc_read,   OPAL_LPC_READ);
+OPAL_CALL(opal_lpc_write,   OPAL_LPC_WRITE);
-- 
1.7.1